simple/multi_mr.c

```diff
@@ -0,0 +1,496 @@
+		g_test_state.mr_state[i].remote->addr = 0;
+
+		memset(g_test_state.mr_state[i].buf, 0, opts.transfer_size);
+
+		g_test_state.mr_state[i].rcntr_next = 1;
+
+		memset(&g_test_state.mr_state[i].remote->key, 0, sizeof(uint64_t));
+
+		g_test_state.mr_state[i].remote->key = i + 1;
+		ret = fi_mr_reg(domain,
+				g_test_state.mr_state[i].buf,
+				opts.transfer_size,
+				FI_REMOTE_WRITE,
+				0,
+				g_test_state.mr_state[i].remote->key,
+				0,
+				&g_test_state.mr_state[i].mr,
+				NULL);
+		if (ret) {
+			FT_PRINTERR("fi_mr_reg", ret);
+			return ret;
+		}
+
+		printf("fi_mr_reg(fi_domain = %p, buffer = %p, memory_region_size = %d, &memory_region = %p, mr_key %lu)\n",
+				domain,
+				g_test_state.mr_state[i].buf,
+				opts.transfer_size,
+				&g_test_state.mr_state[i].mr,
+				g_test_state.mr_state[i].remote->key);
+
+		ret = fi_mr_bind(g_test_state.mr_state[i].mr,
+				&g_test_state.mr_state[i].rcntr->fid,
+				FI_REMOTE_WRITE);
+		if (ret) {
+			FT_PRINTERR("fi_mr_bind", ret);
+			return ret;
+		}
+	}
+
+	return ret;
+}
+
+static int check_buffer(uint8_t *buffer, size_t length, uint8_t pattern)
+{
+	int i;
+
+	for (i = 0; i < length; i++) {
+		if (buffer[i] != pattern) {
+			fprintf(stdout, "ERROR: failed check, %d != %d\n", buffer[i], pattern);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int mr_key_test()
+{
+	int i, ret = 0;
+
+	for (i = 0; i < g_test_state.mr_count; i++) {
+		struct fi_context rma_ctx;
+
+		tx_buf = (char *)g_test_state.mr_state[i].buf;
+
+		if (opts.dst_addr) {
+			memset(g_test_state.mr_state[i].buf,
+					(i + 2), opts.transfer_size);
+
+			printf("write to host's key %lx\n",
+					g_test_state.mr_state[i].remote->key);
+
+			ft_post_rma(FT_RMA_WRITE, ep, opts.transfer_size,
+					g_test_state.mr_state[i].remote, &rma_ctx);
+
+			printf("sent successfully\n");
+
+			ret = wait_cntr(g_test_state.mr_state[i].rcntr,
+					&g_test_state.mr_state[i].rcntr_next);
+			if (ret)
+				return ret;
+
+			printf("check result on memory region with buffer %p and key %lx\n",
+					g_test_state.mr_state[i].buf,
+					g_test_state.mr_state[i].remote->key);
+
+			ret = check_buffer(g_test_state.mr_state[i].buf,
+					opts.transfer_size,
+					(i + 1));
+			if (ret)
+				return ret;
+		} else {
+			ret = wait_cntr(g_test_state.mr_state[i].rcntr,
+					&g_test_state.mr_state[i].rcntr_next);
+			if (ret)
+				return ret;
+
+			printf("check result on memory region with buffer %p and key %lx\n",
+					g_test_state.mr_state[i].buf,
+					g_test_state.mr_state[i].remote->key);
+
+			ret = check_buffer(g_test_state.mr_state[i].buf,
+					opts.transfer_size,
+					(i + 2));
+			if (ret)
+				return ret;
+
+			memset(g_test_state.mr_state[i].buf,
+					(i + 1), opts.transfer_size);
+
+			printf("write to client's key %lx\n",
+					g_test_state.mr_state[i].remote->key);
+
+			ft_post_rma(FT_RMA_WRITE, ep, opts.transfer_size,
+					g_test_state.mr_state[i].remote, &rma_ctx);
+
+			printf("sent successfully\n");
```

Comment:
Limit print output on each iteration, unless some sort of verbose logging is turned on.  I would include the check_buffer() portions only when a data validation command line option has been given.  It's okay to set that option on by default as part of runfabtests.